File: pronacej-backend-seguridad/src/main/java/net/latinus/sistema/integral/gestion/seguridad/service/seguridad/FuncionarioService.java
package net.latinus.sistema.integral.gestion.seguridad.service.seguridad;

import jakarta.servlet.http.HttpServletRequest;
import net.latinus.sistema.integral.gestion.seguridad.model.both.*;
import net.latinus.sistema.integral.gestion.seguridad.model.response.PaginacionResponse;
import net.latinus.sistema.integral.gestion.seguridad.model.response.RespuestaPorDefectoAuditoria;

import java.util.List;

public interface FuncionarioService {

    /**
     * Devuelve una lista de funcionarios
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     * @param bodyEncriptado PaginacionRequest datos de funcionarios.
     *
     * @return RespuestaPorDefectoAuditoria<List<FuncionarioDTO>>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<FuncionarioDTO>> obtenerFuncionarios(HttpServletRequest httpServletRequest,
                                                                                         BodyEncriptado bodyEncriptado);

    /**
     * Devuelve una lista de funcionarios sin paginación
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     *
     * @return RespuestaPorDefectoAuditoria<List<FuncionarioDTO>>
     */
    RespuestaPorDefectoAuditoria<List<FuncionarioDTO>> obtenerFuncionariosSinPaginacion(HttpServletRequest httpServletRequest);
    /**
     * Devuelve una lista de funcionarios filtradas por un parámetro de consulta
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     * @param valor String valor de parámetro de consulta
     * @param bodyEncriptado PaginacionRequest datos de funcionarios de acuerdo a paginador.
     *
     * @return RespuestaPorDefectoAuditoria<List<FuncionarioDTO>>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<FuncionarioDTO>> obtenerFuncionariosPorValor(HttpServletRequest httpServletRequest, String valor,
                                                                                         BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria<FuncionarioDTO> del funcionario creado o actualizado
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     * @param FuncionarioDTO FuncionarioDTO datos del funcionario a crear o actualizar.
     *
     * @return RespuestaPorDefectoAuditoria<FuncionarioDTO>
     */
    RespuestaPorDefectoAuditoria<FuncionarioDTO> crearFuncionario(HttpServletRequest httpServletRequest, FuncionarioDTO FuncionarioDTO);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria<FuncionarioDTO> del funcionario eliminado
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     * @param FuncionarioDTO FuncionarioDTO datos del funcionario a eliminar.
     *
     * @return RespuestaPorDefectoAuditoria<FuncionarioDTO>
     */
    RespuestaPorDefectoAuditoria<FuncionarioDTO> eliminarFuncionario(HttpServletRequest httpServletRequest, FuncionarioDTO FuncionarioDTO);

    /**
     * Devuelve una lista de funcionarios filtradas por un parámetro de consulta
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     *
     * @return RespuestaPorDefectoAuditoria<List<FuncionarioDTO>>
     */
    RespuestaPorDefectoAuditoria<FuncionarioDTO> obtenerFuncionarioDelUsuario(HttpServletRequest httpServletRequest);

    /**
     * Devuelve una lista de jerarquías filtradas por funcionario
     *
     * @param httpServletRequest request peticion.
     *
     * @return RespuestaPorDefectoAuditoria<JerarquiaDTO>
     */
    RespuestaPorDefectoAuditoria<List<JerarquiaDTO>> obtenerJerarquiasPorFuncionarios(HttpServletRequest httpServletRequest);
}

File: pronacej-front/src/app/modules/seguridad/services/funcionario.service.ts
import { Injectable } from '@angular/core';
import { JerarquiaDTO } from 'app/core/model/both/jerarquiaDTO.model';
import { FuncionarioDTO } from 'app/core/model/both/seguridad/FuncionarioDTO.model';
import { PaginacionRequest } from 'app/core/model/request/PaginacionRequest.model';
import { PaginacionResponse } from 'app/core/model/response/PaginacionResponse.model';
import { RespuestaPorDefecto } from 'app/core/model/response/RespuestaPorDefecto.model';
import { BackendService } from 'app/core/services/backend.service';
import { Observable } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class FuncionarioService {
    private path = '/funcionario';

    constructor(private backendService: BackendService) {
        //this.backendService.actualizarClaves();
    }

    obtenerFuncionariosValidos(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<FuncionarioDTO>>> {
        let endPoint = this.path + '/lista';
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    obtenerFuncionariosSinPaginacion(
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<FuncionarioDTO[]>> {
        let endPoint = this.path + '/listaSinPaginacion';
        return this.backendService.getFinal(
            endPoint,
            nemonicoMenu
        );
    }

    crearFuncionario(
        creacionDeFuncionario: FuncionarioDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<FuncionarioDTO>> {
        let endPoint = this.path + '/crear';
        return this.backendService.postFinal(
            endPoint,
            creacionDeFuncionario,
            nemonicoMenu
        );
    }

    eliminarFuncionario(
        creacionDeFuncionario: FuncionarioDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<boolean>> {
        let endPoint = this.path + '/eliminar';
        return this.backendService.postFinal(
            endPoint,
            creacionDeFuncionario,
            nemonicoMenu
        );
    }

    obtenerFuncionarioDelUsuario(
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<FuncionarioDTO>> {
        let endPoint = this.path + '/obtenerFuncionarioDelUsuario';
        return this.backendService.getFinal(endPoint, {}, nemonicoMenu);
    }

    obtenerFuncionariosPorValor(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<FuncionarioDTO>>> {
        let endPoint = this.path + '/buscar';
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    obtenerJerarquiasFuncionarioDelUsuario(
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<JerarquiaDTO[]>> {
        let endPoint = this.path + '/obtenerJerarquiasFuncionarioDelUsuario';
        return this.backendService.getFinal(endPoint, {}, nemonicoMenu);
    }

    async checkError(error: any, mostrarError = true): Promise<string> {
        return await this.backendService.checkError(error, mostrarError);
    }
}
